File: Python/merge-k-sorted-lists.py
import heapq
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return

        h = []
        for i, first_node in enumerate(lists):
            if first_node:
                heapq.heappush(h, (first_node.val, i))

        head = cur = None
        if h:
            value, list_idx = heapq.heappop(h)
            head = cur = lists[list_idx]
            if cur.next:
                heapq.heappush(h, (cur.next.val, list_idx))
                lists[list_idx] = cur.next
        while h:
            value, list_idx = heapq.heappop(h)
            node = lists[list_idx]
            if node.next:
                heapq.heappush(h, (node.next.val, list_idx))
                lists[list_idx] = node.next
            cur.next = node
            cur = node

        return head

File: C++/merge-k-sorted-lists.cpp
#include <vector>
#include <queue>


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        // Time:  O(n * logk)
        // Space: O(k)
        ListNode* dummy = new ListNode(0);

        auto cmp = [](ListNode* lhs, ListNode* rhs) {return lhs->val > rhs->val; }; 
        std::priority_queue<ListNode*, std::vector<ListNode*>, decltype(cmp)> heap(cmp);

        for (const auto& head: lists) {
            if (head) {
                heap.emplace(head);
            }
        }
        ListNode* cur = dummy;
        while (!heap.empty()) {
            ListNode* node = heap.top();
            heap.pop();
            if (node->next) {
                heap.emplace(node->next);
            }
            cur->next = node;
            cur = cur->next;
        }

        return dummy->next;
    }
};
